executor/shuffle.go

```diff
@@ -150,16 +164,27 @@ func (e *ShuffleExec) Close() error {
 		e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, runtimeStats)
 	}
 
-	err := e.dataSource.Close()
-	err1 := e.baseExecutor.Close()
-	if err != nil {
-		return errors.Trace(err)
+	// close dataSources
+	errArr := make([]error, len(e.dataSources))
+	for i, dataSource := range e.dataSources {
+		errArr[i] = dataSource.Close()
 	}
-	return errors.Trace(err1)
+	// check close error
+	for _, err := range errArr {
+		if err != nil {
+			return errors.Trace(err)
```

Comment:
Should we also close `e.baseExecutor` here to avoid resource leaks?
Should we ensure that all children at least be closed once? @huang-b

Reply:
fixed